src/componmnet/PointCloud/render.js

```diff
@@ -49,14 +49,65 @@ const main = (canvas, data, ratio) => {
         nodePool[data[i].id].position.x = data[i].umap1;
         nodePool[data[i].id].position.y = data[i].umap2;
         nodePool[data[i].id].position.z = data[i].umap3;
+        nodePool[data[i].id].userData = {
+            'label': data[i].id,
+        }
         scene.add(nodePool[data[i].id]);
     }
 
     const controls = new OrbitControls(camera, renderer.domElement);
     controls.enablePan = false;
-
     renderer.render(scene, camera);
 
+    // raycasting for showing text
+    const raycaster = new THREE.Raycaster();
+    const mouse = new THREE.Vector3();
+    let mouseX = null;
+    let mouseY = null;
+    let lastLabel = "";
+
+    function onMouseMove(event) {
+        let elem = renderer.domElement;
+        let boundingRect = elem.getBoundingClientRect();
+
+        let x = (event.clientX - boundingRect.left);
+        let y = (event.clientY - boundingRect.top);
+        mouse.x = (x / boundingRect.width) * 2 - 1;
+        mouse.y = - (y / boundingRect.height) * 2 + 1;
+        mouse.z = 0.5;
+        raycaster.setFromCamera(mouse, camera);
+
+        mouseX = event.clientX;
+        mouseY = event.clientY;
+
+    }
+    canvas.addEventListener('mousemove', onMouseMove);
+
+    function render() {
+        const intersects = raycaster.intersectObjects( scene.children );
+
+
+        if (intersects.length > 0) {
+            const curLabel = intersects[0].object.userData['label'];
+            if (curLabel !== lastLabel) {
+                let hovIn = new CustomEvent('clientHovIn', {
+                    detail: {x: mouseX, y: mouseY, label: curLabel}
+                }); 
+                document.dispatchEvent(hovIn);
+                lastLabel = curLabel;
```

Comment:
The two lines could be moved out of the if statement.
Since both if and else statements will do the same thing.

Reply:
The dispatchevent takes two different events so it is different. Also since lastLabel = curLabel; only happens when curLabel != lastLabel I think it should still stay in the statement

Reply:
```suggestion
        let curLabel = '';
        let thisEvent = undefined;
        if (intersects.length > 0) {
            curLabel = intersects[0].object.userData['label'];

            if (curLabel !== lastLabel) {
                thisEvent = new CustomEvent('clientHovIn', {
                    detail: {x: mouseX, y: mouseY, label: curLabel}
                }); 
            }
        } else {
            if (curLabel !== lastLabel) {
                thisEvent = new CustomEvent('clientHovOut', {});
            }
        }

        if (thisEvent) {
            document.dispatchEvent(hovOut);
            lastLabel = curLabel;
        }
        
```

Something like this should work?
Also, shall we change the event name? What does the `client` mean?